fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+//
+// Description:  This file contains general hyperbolic conservation element/face
+//               form integrators.
+//
+//               Abstract RiemannSolver, HyperbolicFormIntegrator, and
+//               DGHyperbolicConservationLaws are defined. Also, several example
+//               hyperbolic equations are defined; including advection, burgers,
+//               shallow water, and euler equations.
+//
+//               To implement a specific hyperbolic conservation laws, users can
+//               create derived classes from HyperbolicFormIntegrator with overloaded
+//               ComputeFlux. One can optionally overload ComputeFluxDotN to avoid
+//               creating dense matrix when computing normal flux.
+//
+//               FormIntegrator use high-order quadrature points to implement
+//               the given form to handle nonlinear flux function. During the
+//               implementation of forms, it updates maximum characteristic
+//               speed and collected by DGHyperbolicConservationLaws. The global
+//               maximum characteristic speed can be obtained by public method
+//               getMaxCharSpeed so that the time step can be determined by CFL
+//               condition. Also, resetMaxCharSpeed should be called after each Mult.
+//
+//               @note For parallel version, users should reduce all maximum
+//               characteristic speed from all processes using MPI_Allreduce.
+//
+// @todo: Implement limiter @dohyun-cse
+// @todo: Implement boundary condition @dohyun-cse
+//
+//
+// Class structure: DGHyperbolicConservationLaws
+//                  |- HyperbolicFormIntegrator: (F(u,x), grad v) and <F̂(u, x, n), [[v]])
+//                  |  |- RiemannSolver: (F±, u±) ↦ F̂(u,x,n)
+//                  |
+//                  |- (Par)NonlinearForm: Evaluate form integrators
+//
+
+#ifndef MFEM_HYPERBOLIC_CONSERVATION_LAWS
+#define MFEM_HYPERBOLIC_CONSERVATION_LAWS
+
+#include "nonlinearform.hpp"
+#ifdef MFEM_USE_MPI
+#include "pnonlinearform.hpp"
+#endif
+namespace mfem
+{
+
+/**
+ * @brief Abstract class for numerical flux for an hyperbolic conservation laws
+ * on a face with states, fluxes and characteristic speed
+ *
+ */
+class RiemannSolver
+{
+public:
+   RiemannSolver() {}
+   /**
+    * @brief Evaluates numerical flux for given states and fluxes. Must be
+    * overloaded in a derived class
+    *
+    * @param[in] state1 state value at a point from the first element
+    * (num_equations x 1)
+    * @param[in] state2 state value at a point from the second element
+    * (num_equations x 1)
+    * @param[in] fluxN1 normal flux value at a point from the first element
+    * (num_equations x dim)
+    * @param[in] fluxN2 normal flux value at a point from the second element
+    * (num_equations x dim)
+    * @param[in] maxE maximum characteristic speed (eigenvalue of flux jacobian)
+    * @param[in] nor normal vector (not a unit vector) (dim x 1)
```

Comment:
```suggestion
    * @param[in] nor normal vector (not a unit vector) (dim)
```

Also the "not a unit vector" is a bit confusing. Is the normal vector expected to be scaled by the face Jacobian determinant?

Reply:
CalcOrtho(const DenseMatrix &J, Vector &nor) gives 'scaled' normal vector. I changed the comment for better understanding.